rollup/da_syncer/l1_message_queue_height_finder.go

```diff
@@ -0,0 +1,127 @@
+package da_syncer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/scroll-tech/go-ethereum/ethdb"
+	"github.com/scroll-tech/go-ethereum/rollup/da_syncer/blob_client"
+	"github.com/scroll-tech/go-ethereum/rollup/da_syncer/da"
+	"github.com/scroll-tech/go-ethereum/rollup/l1"
+)
+
+type L1MessageQueueHeightFinder struct {
+	ctx                context.Context
+	calldataBlobSource *da.CalldataBlobSource
+	l1Reader           *l1.Reader
+}
+
+func NewL1MessageQueueHeightFinder(ctx context.Context, l1height uint64, l1Reader *l1.Reader, blobClient blob_client.BlobClient, db ethdb.Database) (*L1MessageQueueHeightFinder, error) {
+	calldataBlobSource, err := da.NewCalldataBlobSource(ctx, l1height, l1Reader, blobClient, db)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create calldata blob source: %w", err)
+	}
+
+	return &L1MessageQueueHeightFinder{
+		ctx:                ctx,
+		calldataBlobSource: calldataBlobSource,
+		l1Reader:           l1Reader,
+	}, nil
+}
+
+// FindL1MessageQueueHeight finds the L1 message queue height for the target batch.
+// It does so by:
+// 1. find bundle in which target batch was finalized
+// 2. fetch the tx of the bundle to get the height of the L1 message queue after the bundle
+// 3. with this information we can calculate the L1 message count for each batch from last finalized bundle to the target batch.
+func (f *L1MessageQueueHeightFinder) FindL1MessageQueueHeight(targetBatch uint64) (uint64, error) {
+	batches := make(map[uint64]da.EntryWithBlocks)
+
+	// 1. find bundle in which target batch was finalized
+	daEntries, err := f.calldataBlobSource.NextData()
```

Comment:
Why do we expect to find the correct finalize event after one invocation of `NextData` here? Is it because we make sure to pass the correct `l1height1`?

Reply:
you are right, I totally forgot the loop around it :D